### ISSUE TITLE
Resource Prototypes

### ISSUE THREAD

**Issue by scottietremendous** [concourse/concourse#6601]
#### TLDR

- A resource prototype is a prototype that can be used as a resource in a pipeline.

#### Problem

- Under Concourse's existing Resource types implementation, the resource API/contract currently expects resources to only emit a single version on `out`. This is limiting to users who want to write more complicated pipelines where they would want to leverage an array of versions. 
- For instance, if a user does a push of 10 commits, they'd want all commits to be be recorded instead of just the latest one. 

#### Proposed Solution

- Prototypes are configured under `prototypes:` in the pipeline, resource types are under `resource_types:` - this is how we know which protocol/interface to use.
- Resources implemented by prototypes support creating multiple versions from `put` steps.
- Resources implemented by prototypes have an optional `get` after `put` - the user must configure get: and put: on the same step.
- Versions produced by prototypes must be able to be deleted - it may be a good idea to implement #4871 in order to support this more easily.
- A new delete step is added, supported by prototypes but not resource types.

#### Outcomes

- This implementation fills small gaps that were neglected by the non-prototype-based Resource abstraction.
- Introduces a more steady reliable version history - all commits, instead of just the commits since you created the resource.
- Speed and other quality of life improvements

#### Current Status

- Slated to be picked-up in Q4 post-[Prototypes](https://github.com/concourse/concourse/issues/5719) implementation.

#### References

- #5870 - Implement Resource Prototypes
- [RFC: Resource Prototypes](https://github.com/concourse/rfcs/blob/master/038-resource-prototypes/proposal.md)
